src/sources/journald.rs

```diff
@@ -372,32 +297,134 @@ where
                 }
 
                 emit!(JournaldEventReceived {
-                    byte_size: text.len()
+                    byte_size: bytes.len()
                 });
 
-                match channel.send(record).compat().await {
+                match self.out.send(create_event(record)).await {
                     Ok(_) => {}
-                    Err(()) => error!(message = "Could not send journald log."),
+                    Err(error) => {
+                        error!(message = "Could not send journald log.", %error);
+                        // `out` channel is closed, don't restart journalctl.
+                        return false;
+                    }
                 }
             }
 
             if saw_record {
-                if let Some(cursor) = cursor {
-                    if let Err(error) = self.checkpointer.set(&cursor).await {
-                        error!(
-                            message = "Could not set journald checkpoint.",
-                            %error,
-                            filename = ?self.checkpointer.filename,
-                        );
-                    }
-                }
+                Self::save_checkpoint(checkpointer, &*cursor).await;
+            }
+        }
+    }
+
+    async fn save_checkpoint(checkpointer: &mut Checkpointer, cursor: &Option<String>) {
+        if let Some(cursor) = cursor {
+            if let Err(error) = checkpointer.set(cursor).await {
+                error!(
+                    message = "Could not set journald checkpoint.",
+                    %error,
+                    filename = ?checkpointer.filename,
+                );
             }
         }
     }
 }
 
-fn decode_record(text: &str, remap: bool) -> Result<Record, JsonError> {
-    let mut record = serde_json::from_str::<JsonValue>(&text)?;
+/// A function that starts journalctl process.
+/// Return a stream of output splitted by '\n', and a `StopJournalctlFn`.
+///
+/// Code uses `start_journalctl` below,
+/// but we need this type to implement fake journald source in testing.
+type StartJournalctlFn = Box<
+    dyn Fn(
+            &Option<String>, // cursor
+        ) -> crate::Result<(BoxStream<'static, io::Result<Bytes>>, StopJournalctlFn)>
+        + Send
+        + Sync,
+>;
```

Comment:
What are the advantages of using this boxed function over the trait type it appears to be replacing? Could this be accomplished using a trait, which is (arguably) more idiomatic for Rust and Vector?

Reply:
Previously, it returns `Self` and `Self` needs to implement `Stream`, I found that hard to follow and would change it to return a BoxStream if I keep the trait.

Personally, I assume more than one type are being used when I see a trait or a generic type, but there was only one type/function here (except in testing). And to me, this looks more like a function than an object/interface, after all it was a trait with only one function.